python/graphstorm/gconstruct/transform.py

```diff
@@ -935,6 +910,149 @@ def call(self, feats):
+
+    Parameters
+    ----------
+    col_name : str
+        The name of the column that contains the feature.
+    feat_name : str
+        The feature name used in the constructed graph.
+    separator : str
+        The separator to split data into multiple node ids.
+    """
+    def __init__(self, col_name, feat_name, separator=None):
+        self._target_ntype = None
+        self._target_etype = None
+        self._nid_map = None
+        self._separator = separator
+        super().__init__(col_name, feat_name, out_dtype=np.int64)
+
+    def set_target_etype(self, etype):
+        """ Set the etype of this hard edge negative transformation ops
+            and associated hard negative information. For example,
+            self._target_ntype.
+
+        Parameters
+        ----------
+        etype : tuple of str
+            The edge type the hard negatives belonging to.
+        """
+        raise NotImplementedError
+
+    @property
+    def target_etype(self):
+        """ The the edge type of this hard negative transformation.
+        """
+        return self._target_etype
+
+    @property
+    def neg_ntype(self):
+        """ Return the node type of hard negatives
+        """
+        return self._target_ntype
+
+    def set_id_maps(self, id_maps):
+        """ Set ID mapping for converting raw string ID to Graph ID
+        """
+        assert self._target_ntype is not None, \
+            "The target node type should be set, it can be the source node type " \
+            "or the destination node type depending on the hard negative case."
+        assert self._target_ntype in id_maps, \
+            f"The nid mapping should have the mapping for {self._target_ntype}. " \
+            f"But only has {id_maps.keys()}"
+        self._nid_map = id_maps
+
+    def pre_process(self, feats):
+        """ Pre-process input feats
+
+            Not all the edges have the same number of hard negatives.
+            Thus we need to know the maxmun number of hard negatives first.
+
+        Parameters
+        ----------
+        feats:
+            feats to be processed
+
+        Return:
+            dict: {feature_name: feats_statistics}
+        """
+        assert isinstance(feats, (np.ndarray, ExtMemArrayWrapper)), \
+            "Feature of HardEdgeNegativeTransform must be numpy array or ExtMemArray"
+
+        if self._separator is None:
+            max_dim = feats.shape[1]
+        else:
+            assert len(feats.shape) == 1 or feats.shape[1] == 1, \
+                "When a separator is given, the input feats must be a list of strings."
+
+
+            if feats.dtype.type is not np.str_:
+                logging.warning("When a separator is given, "
+                                "the input feats must be a list of strings."
+                                "But we get %s. Will convert the input feats into str array",
+                                feats.dtype.type)
```

Comment:
is it better to assert here? if the input isn't string, even if we convert the dtype, does it still work?

Reply:
We can not assert it. If we load data from parquet, the dtype will be np.object. Thus, we need to do the conversion. This is a hint for user that the data stored should be string.

Reply:
if so, should we print warning? it seems that it's a normal behavior that dtype is not string.

Reply:
make sense